packages/nextjs/hooks/scaffold-eth/useScaffoldWriteContract.ts

```diff
@@ -15,23 +15,56 @@ import {
   UseScaffoldWriteConfig,
 } from "~~/utils/scaffold-eth/contract";
 
+type ScaffoldWriteContractReturnType<TContractName extends ContractName> = Omit<
+  ReturnType<typeof useWriteContract>,
+  "writeContract" | "writeContractAsync"
+> & {
+  isMining: boolean;
+  writeContractAsync: <
+    TFunctionName extends ExtractAbiFunctionNames<ContractAbi<TContractName>, "nonpayable" | "payable">,
+  >(
+    variables: ScaffoldWriteContractVariables<TContractName, TFunctionName>,
+    options?: ScaffoldWriteContractOptions,
+  ) => Promise<WriteContractReturnType | undefined>;
+  writeContract: <TFunctionName extends ExtractAbiFunctionNames<ContractAbi<TContractName>, "nonpayable" | "payable">>(
+    variables: ScaffoldWriteContractVariables<TContractName, TFunctionName>,
+    options?: Omit<ScaffoldWriteContractOptions, "onBlockConfirmation" | "blockConfirmations">,
+  ) => void;
+};
+
+export function useScaffoldWriteContract<TContractName extends ContractName>(
+  config: UseScaffoldWriteConfig<TContractName>,
+): ScaffoldWriteContractReturnType<TContractName>;
+/**
+ * @deprecated Use object parameter version instead: useScaffoldWriteContract({ contractName: "YourContract" })
+ */
+export function useScaffoldWriteContract<TContractName extends ContractName>(
+  contractName: TContractName,
+  writeContractParams?: UseWriteContractParameters,
+): ScaffoldWriteContractReturnType<TContractName>;
+
 /**
  * Wrapper around wagmi's useWriteContract hook which automatically loads (by name) the contract ABI and address from
  * the contracts present in deployedContracts.ts & externalContracts.ts corresponding to targetNetworks configured in scaffold.config.ts
  * @param contractName - name of the contract to be written to
  * @param config.chainId - optional chainId that is configured with the scaffold project to make use for multi-chain interactions.
  * @param writeContractParams - wagmi's useWriteContract parameters
  */
-export const useScaffoldWriteContract = <TContractName extends ContractName>({
-  contractName,
-  chainId,
-  writeContractParams,
-}: UseScaffoldWriteConfig<TContractName>) => {
+export function useScaffoldWriteContract<TContractName extends ContractName>(
+  configOrName: UseScaffoldWriteConfig<TContractName> | TContractName,
+  writeContractParams?: UseWriteContractParameters,
+): ScaffoldWriteContractReturnType<TContractName> {
+  const finalConfig: UseScaffoldWriteConfig<TContractName> =
+    typeof configOrName === "string"
+      ? { contractName: configOrName, writeContractParams, chainId: undefined }
+      : (configOrName as any);
```

Comment:
Do we need to specify the type here? It seems that it can be inferred.

If not, should it be `UseScaffoldWriteConfig<TContractName>`?

Reply:
If we remove the its being inffered as `any` lol it's fine but in next step when I was destructuring I wanted that autcompletions with keys. 

```suggestion
  const finalConfig =
    typeof configOrName === "string"
      ? { contractName: configOrName, writeContractParams, chainId: undefined }
      : configOrName;
``` 

Actually earlier while playing I had this. 

Inferring  + not asserting `any` at `configOrName`. 

But the problem is when `deployedContracts.ts` is empty. the type `TContractName` i.e `export type ContractName = keyof Contracts;` is inferred as `string | number` instead of `ContractName` being only `string`  [weird how TS works](https://stackoverflow.com/questions/57185972/typescript-infers-object-key-to-number-instead-of-string). 

Now since it is inferred  as `string | number` in the next step I wasn't able to destructure  it becoz it was giving error that `contractName, writeContractParams` doesn't exist on type `string | number`.

Reply:
Damn, types.

Thanks for the explanation!! <3

Reply:
> If not, should it be UseScaffoldWriteConfig<TContractName>?

Yeah just asserted the `configOrName` with it and now its being inferred at https://github.com/scaffold-eth/scaffold-eth-2/pull/1015/commits/3bef6823cc89805fce3b67e974584fa52006a28f

lol forgot to click "add comment" here